### ISSUE TITLE
type species

### ISSUE THREAD

**Issue by touretzkyds** [touretzkyds/KnowledgeGraphDemo#18]
Some family and genus taxa have a "type species" property in WikiData. Let's import this property into our own knowledge graph. If we don't already have the type species included in our graph,  then please add that species.

**Comment by deyaliao** [comment]
Hi, I'm confused about how exactly to implement this––although the "genus" and "family" webpages have the property type species, individual taxons don't have any type species information on their web page. For example if we go to the original type species [webpage](https://en.wikipedia.org/wiki/Type_species), on the right it shows an example: _cygnus cygnus_ is the type species for the species of the genus _cygnus_. However, if we go to the [cygnus](https://www.wikidata.org/wiki/Q29564928) or [cygnus cygnus](https://www.wikidata.org/wiki/Q25612) page, there is nothing to indicate that there is any relation with type species. Additionally, I've checked for many examples of taxons (of type rank family or genus), and not a single one has any information about type species. Should I continue looking for a way in WikiData, or perhaps use some other external database––such as the reptile one you linked last week?

**Comment by touretzkyds** [comment]
This seems to be another example of WikiData weirdness. I can't find any type species designations in WikiData either, although there is a good deal of information about the concept of "type species":
````
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
SELECT ?r ?s ?z WHERE {
  wd:Q252730 ?r [?s ?z] filter ((?s != rdfs:label) && (?s != schema:description) && (?s != skos:altLabel))
  } ORDER BY ?r
````
If we can get type species info from ITIS or the reptile database, then let's define kgo:typeSpecies and add this info to our knowledge graph.

**Comment by deyaliao** [comment]
ITIS doesn't seem to have any type species information, but I did download the reptile database to scrape its type species information. The reptile database organized its information quite poorly in my opinion, but I ended up with a final dictionary of information (all code is [here](https://github.com/deyaliao/knowledgeGraphs/tree/master/reptiles-type-species), with the specific type species dictionary at the very bottom [here](https://github.com/deyaliao/knowledgeGraphs/blob/master/reptiles-type-species/reptile-database.py)), where the genus was the key, and the type species was its value. I copy/pasted this dictionary into my general algorithm; for any taxa that I came across, I checked if there was a type species associated with it––but of course this is kind of limiting, since the only type species information we have on hand at the moment is with regards to reptiles. 

Currently, kgo:typeSpecies is of type rdfs:Literal (just the species in plain English) instead of kgo:Taxon (the Q-value of the species) just simply because species are processed (I just insert the Q-value with regards to the species into a dictionary for future access) after genuses, meaning the Q-values in regards to each species aren’t available at the time I access the genus’s type species. Perhaps, I could replace the Q-value of the type species at the very end, but that seems to be a bit difficult, so any suggestions are appreciated. 

Additionally, how I organize the information right now is that for all GENUS with type species, there is information to indicate kgo:typeSpecies “type-species”. But, for the individual SPECIES that are the genus’s dictated type species, I currently don’t have any information to indicate that this species is indeed the type species. For example, for the Genus “Iguana”, the triples look like:

```
boltz:Q205531 a kgo:Taxon ;
   kgo:subTaxonOf  boltz:Q21400931 ;
   kgo:taxonName   "Iguana"@en ;
   kgo:taxonRank   boltz:Q34740 ;
   kgo:typeSpecies "Iguana iguana"@en ;
   kgo:taxonImage  <http://commons.wikimedia.org/wiki/Special:FilePath/Portrait%20of%20an%20Iguana.jpg> ;
   skos:definition """Genus of reptiles"""@en ;
   skos:prefLabel  "Iguana"@en .
```
But, for the typeSpecies “iguana iguana”, there’s no definition to show that it’s the designated typeSpecies. Should we define a new property? I don’t think it makes sense to use kgo:typeSpecies, as the definition of that right now is “The taxon has type species of the object”. 

Regardless, the type species information is in our knowledge graph for the (limited reptile) genuses. We can access our limited reptile species through this query (essentially, of the 100 I imported last week, only 2 genus level taxa happened to match the ones included in the reptile database, which makes sense since the 100 includes a wide variety of animals): 

```
PREFIX kgo: <http://solid.boltz.cs.cmu.edu:3030/ontology/>
PREFIX rdfs:  <http://www.w3.org/2000/01/rdf-schema#> 

  SELECT DISTINCT ?yLabel ?rLabel ?typeS WHERE {
    ?a kgo:taxonName "Animalia"@en.
    ?y kgo:subTaxonOf* ?a.
    ?y kgo:taxonName ?yLabel.
    ?y kgo:typeSpecies ?typeS.
  ?y kgo:taxonRank [rdfs:label ?rLabel] .  
  }
  ORDER BY ?yLabel
```

Tomorrow, I’m planning to import all 146 reptile species in the database (that happen to be type species of genus present in the reptile database), along with the 300 primate species you assigned to me in a different issue, just so we have more “type species” information in the Knowledge Graph. I’ll also look for other databases that contain type-species information so our knowledge graph can be even more expansive.

**Comment by touretzkyds** [comment]
Too bad ITIS lacks type species info. I agree with your decision to get the data from the reptile database and add in the reptile genuses and their type species. And looking for similar sources for birds, mammals, etc. is a good idea.

We can add a kgo:typeSpeciesOf link to point back to the family or genus. This makes the graph more loopy and less tree-like, which at this point is a good thing to help us explore presentation and browsing styles. We can always remove the link if we change our minds later.

**Comment by deyaliao** [comment]
I added the kgo:typeSpeciesOf property, so this is now available in our database. There are more results with regards to both kgo:typeSpecies and kgo:typeSpeciesOf, as I imported the species from 3 reptile orders. This query specifically shows the species that are typeSpecies of a genus:

```
PREFIX kgo: <http://solid.boltz.cs.cmu.edu:3030/ontology/>
PREFIX rdfs:  <http://www.w3.org/2000/01/rdf-schema#> 

  SELECT DISTINCT ?yLabel ?rLabel ?typeS  ?typeSof WHERE {
    ?a kgo:taxonName "Animalia"@en.
    ?y kgo:subTaxonOf* ?a.
    ?y kgo:taxonName ?yLabel.
  #?y kgo:typeSpecies ?typeS.
  ?y kgo:typeSpeciesOf ?typeSof.
  	?y kgo:taxonRank [rdfs:label ?rLabel] .  
  }
  ORDER BY ?yLabel
```
 

Back to WikiData and typeSpecies information, I’m just a bit confused because if we look at an arbitrary genus such as [Bison’s Wikipedia Page](https://en.wikipedia.org/wiki/Bison), on the right side, they have “Type Species” information that explicitly says that its Type Species is Bos Bison. However, if we look at the WikiData page with regards to [Bison](https://www.wikidata.org/wiki/Q18099), there’s simply no information about type species. This is true for many, many (more well known) genuses (I pulled random genus from ITIS): the [Swan](https://en.wikipedia.org/wiki/Swan), the [Nycteris](https://en.wikipedia.org/wiki/Nycteris) (a bat), and the simple [Alligator](https://en.wikipedia.org/wiki/Alligator). 

But, I found this website http://www.animalbase.org/ that seems to include type species information. If we search cygnus, alligator, bison, etc, all have that info. Looking at their about section, they seem to have a lot of taxa, so that’s quite helpful for us.

**Comment by touretzkyds** [comment]
I guess there's no guarantee that Wikipedia and WikiData are mutually consistent. I'm not surprised that there is info in one that is missing in the other.

I looked up "Macaca" on animalbase.org and found the type species was Simia Inuus, which is strange. Turns out Simia is a now outdated taxonomic term. Simia inuus refers to the Barbary Macaque; see here: https://en.wikipedia.org/wiki/Simia
and also here: https://www.mindat.org/taxon-7373308.html

The modern name for the Barbary macaque is Macaca sylvanus. Wikipedia lists a lot of macaque species at https://en.wikipedia.org/wiki/Macaque but we only have four in our knowledge graph, and Macaca sylvanus is not one of them. Why are we missing the rest?

**Comment by deyaliao** [comment]
With regards to animal base, I'm also not super surprised that there are some outdated terms. For the reptile database, there was some discrepancies too. For example, it says that [Amphisbaena prunicolor](https://reptile-database.reptarium.cz/species?genus=Amphisbaena&species=prunicolor&search_param=%28%28search%3D%27Aporarchus%27%29%29) is the type species of the genus Aporarchus and not Amphisbaena, but there is no existence of a genus Aporarchus on ITIS. I later searched on Wikipedia that they are synonyms for each other, but 1) ITIS has a synonym feature and made no mention of this, their sentence 2) "This species is the type species of the genus Aporarchus which is distinguished from Amphisbaena only by the lack of precloacal pores" made it sound like they were 2 separate genuses? Anyways, are we ruling out animal base?

With regards to the macaque species, I looked back at the code, and found that while all the macaca species are there, it seems like many don't have the subtaxonOf property that indicates it's a subtaxon of Macaca, which might of been what you used to search for all the species under the Genus Macaca. This is because in the [primates csv file](https://github.com/deyaliao/knowledgeGraphs/blob/master/primates.csv) (you can't seem to see the contents on Github, but if you'd like to download it and view it on your desktop), for some reason, all the taxa are not ranked by taxonomical hierarchy, so there are a lot of Macaca species that are listed before the definition of the Macaca genus itself. This is important because previously, I was taking for granted that for each taxa in the file, I would've seen the parent before, so I directly searched a dictionary where I kept track of the taxonID (on the ITIS database) and its corresponding Q-value––Arthur and I came up with this together. This worked because I previously downloaded each species one at a time, where everything was ranked in order of hierarchy. 

I would have to sort the data in the order of taxonomical hierarchy to continue doing it this way (make sure I process all genus before species). Or, I can directly search the dataframe for the row marked by the parent identifier (ex. macaca sylvanus's parenttaxon ID is 180097, so I'll just use pandas functions to access the row where the ID is that number), and from that row, grab the taxonName, and just run an extra query in WikiData with that name––in the case that we reach a species where we haven't processed its genus yet. Which one do you think is faster in runtime?

**Comment by touretzkyds** [comment]
Increasing the number of WikiData network queries is surely much slower than any computation that can be done in local memory.

My suggestion would be to use a two-pass algorithm. The first pass loads in all the data so you know the names and ids of all taxons. The second pass writes out the ttl entries, generating subTaxonOf links as you go. This way you're not dependent on the source data being in any particular order. It also makes it easier to combine data from multiple sources if that becomes necessary.

**Comment by touretzkyds** [comment]
Animalbase still looks like a valuable resource, so I don't want to rule it out. But when there are conflicts with ITIS, let's go with ITIS for now. Also that mindat.org site looked interesting. Can you poke around over there and let me know what you think of it?

**Comment by deyaliao** [comment]
I implemented the two-pass algorithm: if we query again, all the species should show up. 
```
PREFIX kgo: <http://solid.boltz.cs.cmu.edu:3030/ontology/>
PREFIX rdfs:  <http://www.w3.org/2000/01/rdf-schema#> 

  SELECT DISTINCT ?yLabel ?rLabel ?typeS  ?typeSof WHERE {
    ?a kgo:taxonName "Macaca"@en.
    ?y kgo:subTaxonOf* ?a.
    ?y kgo:taxonName ?yLabel.
  OPTIONAL{?y kgo:typeSpecies ?typeS.}
  OPTIONAL{?y kgo:typeSpeciesOf ?typeSof.}
  	?y kgo:taxonRank [rdfs:label ?rLabel] .  
  }
  ORDER BY ?yLabel
```
 The mindat.org site does have a lot of information on it; each mineral especially seems to have a lot of properties available. I will say navigation isn't as intuitive as the ITIS, but that is something I can definitely work with. I will say their rock pages aren't as developed (and load really slowly, on my end), for example [this page](https://www.mindat.org/min-50859.html). Also, they have this in their FAQs: 

> **I have a genuine reason that I'd like to use your data in my project, can I?**
> 
> Please talk to us - we're happy to try to help people use our data in a responsible way. Note that we are unlikely to agree to something that allows significant portions of the database to be copied out of the site unless they are linked in such a way that updates to mistakes on mindat.org automatically update and change the copy that you are using, for example using an XML feed. We are also restricted in what we can and cannot allow permission for, see below.
> 
> In certain cases we will build interfaces for you to help you use our data. For example, the mineral list within the Wikipedia online encyclopedia is generated automatically from a list provided by mindat.org. This list of data is provided under the standard Wikipedia licence for use without any further restrictions.

Also, this is kind of unrelated, but when downloading data from ITIS, I get a folder with three files: the meta.xml file, another xml file related to the data, and the txt data with all the taxon information (this is what I'm using now). This is an [example](https://github.com/deyaliao/knowledgeGraphs/tree/master/Anthozoa%20(example)); I was just wondering if they were at all useful?

**Comment by touretzkyds** [comment]
These xml files don't look particularly useful. One alerts us to all the properties they are encoding, but I don't think we want to mirror all the detail in the ITIS records. We're good with what we have for now.

The mindat.org database is interesting. It's experimental so you can't even get to it from their home page. I found it accidentaly through a Google search. They're interested in fossils, so they have a lot of taxa for extinct species.

**Comment by deyaliao** [comment]
I noticed that my algorithm parsed using “Type Species: “, revealing 146 type species, but in reality, the majority of the type species in the database are defined using “Type species: “, with a lowercase ‘s’ in species. After fixing this, I was able to scrape 987 total type species. Out of these type species & genus combinations, a lot of them were using invalid species names, and I had to manually change a lot of the names to reflect the ITIS counterpart, such that there would be a valid Q-value associated with the taxon. 

I also changed the property to be of boltz:Q… instead of its English label, so this should be reflected in the Data.ttl file. I already updated the server, so if we run this algorithm:
```
PREFIX kgo: <http://solid.boltz.cs.cmu.edu:3030/ontology/>
PREFIX rdfs:  <http://www.w3.org/2000/01/rdf-schema#> 
PREFIX skos: <http://www.w3.org/2004/02/skos/core#>

  SELECT DISTINCT ?a ?y ?yName ?yLabel ?rLabel ?definition ?typeS ?typeSof WHERE {
    ?a kgo:taxonName "Animalia"@en.
    ?y kgo:subTaxonOf* ?a.
    ?y kgo:taxonName ?yName .
  	?y skos:prefLabel ?yLabel .
 	?y skos:definition ?definition .
  ?y kgo:typeSpecies  ?typeS.
  OPTIONAL{?y kgo:typeSpeciesOf [rdfs:label ?typeSof].}
 	?y kgo:taxonRank [rdfs:label ?rLabel] .  
 }
  ORDER BY ?yLabel
```
 
We can see that typeSpecies and typeSpecies are of q-labels concepts.

**Comment by touretzkyds** [comment]
The line "?y kgo:typeSpeciesOf [rdfs:label ?typeSof]" makes no sense as it would require ?y to both have a type species and be a type species.

Here's a query that does work:
````
PREFIX kgo: <http://solid.boltz.cs.cmu.edu:3030/ontology/>
PREFIX rdfs:  <http://www.w3.org/2000/01/rdf-schema#>
PREFIX skos: <http://www.w3.org/2004/02/skos/core#>

SELECT DISTINCT ?a ?y ?yName ?yLabel ?rLabel ?z ?zLabel WHERE {
    ?a kgo:taxonName "Animalia"@en.
    ?y kgo:subTaxonOf* ?a.
    ?y kgo:taxonName ?yName .
    ?y skos:prefLabel ?yLabel .
    ?y kgo:typeSpecies  ?z.
   OPTIONAL {?z kgo:taxonName ?zLabel.}
    ?y kgo:taxonRank [rdfs:label ?rLabel] .
 }
  ORDER BY ?yLabel
````
This turns up some anomalies in the graph. About 10 of the 83 results have a kgo:typeSpecies that is a string rather than a concept. Also, why do the taxa have only a skos:prefLabel and not an rdfs:label? I think every concept should have an rdfs:label.

Also, the Data.ttl file contains 72 typeSpeciesOf entries but 82 typeSpecies entries. Are we missing 10 entries? And when I run the query above I get 83 results, not 82. What's missing?

**Comment by deyaliao** [comment]
Oh, the 10 of the 82 results are my bad; there's one species that is spelled a bit differently than the ITIS database ones, and some of the other genuses have typeSpecies that simply aren't within our database. For example, I downloaded the data file for [this species](https://www.itis.gov/servlet/SingleRpt/SingleRpt?search_topic=TSN&search_value=1125439#null), which has information about the genus Amphisbaena––which has a type species of Amphisbaena fuliginosa, but I never imported this species anytime. 

To elaborate on why my algorithm doesn't work, I use the typeSpecies name as a key to retrieve the corresponding Q-value, but since these keys don't really exist in my dictionary, I can't retrieve the matching Q-value and I put its name as a placeholder for now. This corresponds directly to 10 less typeSpeciesOf entries too; the type species that the reptile dictionary (imported from the reptile database) refers to for that genus is simply not a concept in the ITIS taxa that I imported, hence 10 less. Let me know if that explanation isn't too clear, but both should be fixed now (both queries return 82 results). I changed the wrong spelling, and I just downloaded the other 9 species off of ITIS. 

I'm not too sure why it's skos:prefLabel over rdfs:label, but I just know that it was always like that? If you look at Arthur's original taxa in [Data.ttl](https://github.com/touretzkyds/KnowledgeGraphServer/blob/ebcff7ccbea4b34ed9a3d8746763c05abff8d5c0/devel/Data.ttl#L557), they all use skos:prefLabel over rdfs:label, so I modeled my triples after that. I can change it to rdfs:label easily, however? Let me know if that's what you want.

**Comment by touretzkyds** [comment]
My thinking on this issue is evolving. What I think now is that every concept should have an rdfs:label that matches the concept name. For taxa this should be the official taxon name, e.g. "Animalia". But we can optionally add skos:prefLabel and even skos:altLabel to give things like common names, nicknames, etc. So for Q729 the rdfs:label would be "Animalia" but the skos:prefLabel should be "Animal". This is different from how WikIData does it; they use the common name as the rdfs:label and the official taxonomic name as one of several skos:altLabel values.

**Comment by deyaliao** [comment]
Sure, so should I automatically give all taxa an rdfs:label matching the concept name, and if they have another label that WikiData uses as the label, such as "bird" with "aves", then I use that as skos:prefLabel? In other words, skos:prefLabel is always the common name?

**Comment by touretzkyds** [comment]
Yes, let's do that.

**Comment by deyaliao** [comment]
I changed this, but because not every taxa has a skos:prefLabel entry, the demo2d is broken as of right now, since the algorithm attempts to get that information from every taxon I believe. I'm still not too familiar with the demo page loading algorithm, so I'll take some more time to fix that today.

**Comment by touretzkyds** [comment]
demo2d is not just broken, it's broken so badly that it infinite loops and the tab cannot be closed. This should _never_ happen. We need to write code defensively so that it fails gracefully when it gets bad data.

**Comment by deyaliao** [comment]
Yes, I agree! Reapproaching how we fetch the data is a good idea. Perhaps we can talk about this during our meeting.
